runtime/lua/vim/lsp/util.lua

```diff
@@ -1484,6 +1479,44 @@ function M.stylize_markdown(bufnr, contents, opts)
   return stripped
 end
 
+--- Removes surrounding empty lines in the given markdown content, replaces
+--- separators by horizontal dividers spanning the entire window, and configures
+--- the given buffer.
+---
+---@param contents string[] of lines to show in window
+---@param opts table with optional fields
+---  - height    of floating window
+---  - width     of floating window
+---  - wrap_at   character to wrap at for computing height
+---  - max_width  maximal width of floating window
+---  - max_height maximal height of floating window
+--- @return string[] contents with the processed lines
+function M.normalize_markdown(bufnr, contents, opts)
+  validate({
+    contents = { contents, 't' },
+    opts = { opts, 't', true },
+  })
+  opts = opts or {}
+
+  -- Clean up
+  contents = M.trim_empty_lines(contents)
+
+  -- Remove carriage returns
+  contents = vim.split(table.concat(contents, '\n'):gsub('\r', ''), '\n')
+
+  -- Compute size of float needed to show (wrapped) lines
+  opts.wrap_at = opts.wrap_at or (vim.wo['wrap'] and api.nvim_win_get_width(0))
+  local width = M._make_floating_popup_size(contents, opts)
+
+  -- Replace separators
+  local divider = string.rep('─', math.min(width, opts.wrap_at or width))
+  contents = replace_separators(contents, divider)
+
+  api.nvim_buf_set_lines(bufnr, 0, -1, false, contents)
+
+  return contents
+end
```

Comment:
This function should not have any side effects, it should be a simple "Markdown goes in, Markdown comes out" function.

```suggestion
--- Normalizes Markdown input to a canonical form.
---
--- The returned Markdown adheres to the GitHub Flavored Markdown (GFM)
--- specification.
---
--- The following transformations are made:
---
---   1. Empty lines at the beginning or end of the content are removed
---   2. Carriage returns ('\r') are removed
---   3. Successive empty lines are collapsed into a single empty line
---   4. Thematic breaks are expanded to the given width
---
---@param contents string[] Table of lines containing valid Markdown
---@param opts table Optional arguments:
---  - width (integer): Thematic breaks are expanded to this size. Defaults to 80.
---@return string[] Table of lines containing normalized Markdown
---@see https://github.github.com/gfm
function M.normalize_markdown(contents, opts)
  validate({
    contents = { contents, 't' },
    opts = { opts, 't', true },
  })

  opts = opts or {}

  -- 1. Empty lines at the beginning or end of the content are removed
  contents = M.trim_empty_lines(contents)

  -- 2. Carriage returns are removed
  contents = vim.split(table.concat(contents, '\n'):gsub('\r', ''), '\n')

  -- 3. Successive empty lines are collapsed into a single empty line
  -- TODO?

  -- 4. Thematic breaks are expanded to the given width
  local divider = string.rep('─', math.min(opts.width or 80))
  contents = replace_separators(contents, divider)

  return contents
end

```

Reply:
If it gets this feature out the door sooner then we can leave this as-is for now, but this change should be made in a follow up then.

This function will be moved out of `vim.lsp.util` and when it does, it's important that it is a simple transformation function.

Reply:
@gpanders I will apply your suggestions and update the call sites to set the lines in the buffer. I won't be moving the function to a new module yet though since I believe we haven't even agreed on where to put it yet.